include/geom/elem_side_builder.h

```diff
@@ -0,0 +1,49 @@
+// The libMesh Finite Element Library.
+// Copyright (C) 2002-2021 Benjamin S. Kirk, John W. Peterson, Roy H. Stogner
+
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+
+
+
+#ifndef LIBMESH_ELEM_SIDE_BUILDER_H
+#define LIBMESH_ELEM_SIDE_BUILDER_H
+
+#include "libmesh/elem.h"
+
+namespace libMesh
+{
+
+/**
+ * Helper for building element sides that minimizes the construction
+ * of new elements.
+ *
+ * TODO: add more info
+ */
+class ElemSideBuilder
+{
+public:
+  /**
+   * Builds an element side for side \p s of element \p elem.
+   */
+  Elem & operator(const Elem & elem, const unsigned int s);
+
+private:
+  /// Proxy element cache for building side pointers; indexed by ElemType
+  std::vector<std::unique_ptr<Elem>> _cached_elems;
```

Comment:
This is going to end up quite sparse, isn't it?  I can't think of a better way to do it, though; typically for containers this small the relative overhead of anything but a vector/array is huge.

Reply:
Yep. Only other option here is if we want to go static with an array; same sparsity issue though